plugins/inventory/vmware_vm_inventory.py

```diff
@@ -395,211 +485,255 @@ def parse(self, inventory, loader, path, cache=True):
+                    p_name = self._sanitize_group_name('/'.join(parents))
+                    p_group = self.inventory.add_group(p_name)
+
+                    self.inventory.add_child(p_group, c_group)
+                    c_group = p_group
+                    parents.pop()
+
+        # Hostvars formats manipulation
+        host_properties = host_properties if with_nested_properties else to_flatten_dict(host_properties)
+
+        if self.get_option('with_sanitized_property_name'):
+            host_properties = camel_dict_to_snake_dict(host_properties)
+
+        can_sanitize = self.get_option('with_sanitized_property_name')
+        for k, v in host_properties.items():
+            k = self._sanitize_group_name(k) if can_sanitize else k
+            self.inventory.set_variable(host, k, v)
+
+
+def parse_vim_property(vim_prop, key=""):
+    # For '--yaml' sake!
+    #   Unexpected Exception, this is probably a bug: ('cannot represent an object', <data>
+    prop_type = type(vim_prop).__name__
+    if prop_type.startswith("vim") or prop_type.startswith("vmodl"):
+        if isinstance(vim_prop, DataObject):
+            r = {}
+            for prop in vim_prop._GetPropertyList():
+                if prop.name not in ['dynamicProperty', 'dynamicType', 'managedObjectType']:
+                    sub_prop = getattr(vim_prop, prop.name)
+                    r[prop.name] = parse_vim_property(sub_prop, prop.name)
+            return r
+
+        elif isinstance(vim_prop, list):
+            r = []
+            for prop in vim_prop:
+                r.append(parse_vim_property(prop))
+            return r
+        return vim_prop.__str__()
+
+    elif prop_type == "datetime":
+        return Iso8601.ISO8601Format(vim_prop)
+
+    elif prop_type == "long":
+        return int(vim_prop)
+    elif prop_type == "long[]":
+        return [int(x) for x in vim_prop]
+
+    elif isinstance(vim_prop, list):
+        return [parse_vim_property(x) for x in vim_prop]
+
+    elif prop_type in ['bool', 'int', 'NoneType']:
+        return vim_prop
+
+    return to_text(vim_prop)
+
+
+def to_nested_dict(vm_properties):
+    host_properties = {}
+
+    for vm_prop_name, vm_prop_val in vm_properties.items():
+        prop_parents = reversed(vm_prop_name.split("."))
+        prop_dict = parse_vim_property(vm_prop_val, vm_prop_name)
+
+        for k in prop_parents:
+            prop_dict = {k: prop_dict}
+        host_properties = dict_merge(host_properties, prop_dict)
+
+    return host_properties
+
+
+def to_flatten_dict(d, parent_key='', sep='.'):
+    items = []
+    for k, v in d.items():
+        new_key = parent_key + sep + k if parent_key else k
+        if v and isinstance(v, dict):
+            items.extend(to_flatten_dict(v, new_key, sep=sep).items())
+        else:
+            items.append((new_key, v))
+    return dict(items)
+
+
+# Patch AnsibleDumper
+# This for 'ansible-test' with '--debug'
+# TODO: Move it to 'ansible.parsing.yaml.dumper'
```

Comment:
Move to Ansible base? That is kind of a bold TODO. I'd like to get that vetted a little bit.

Looking into `represent_hostvars`

https://github.com/ansible/ansible/blob/308723c3ca1122363e419070f1fa1d76ff5611a9/lib/ansible/parsing/yaml/dumper.py#L38

```python
def represent_hostvars(self, data):
    return self.represent_dict(dict(data))
```

This is a deceptively simple thing - I'm guessing that it allows serializing python objects which can be coerced to dicts, but are not dicts themselves. Serializing python objects is a common headache for `ansible-inventory` when the inventory plugin uses some client library that has its own native types.

Reply:
`represent_hostvars` is already used for `HostVars` and `HostVarsVars` but not for `VarsWithSources`. Since yaml loader is used in safe mode it limits it ability to a number of of built-in python objects. I believed that `VarsWithSources` fall through the cracks and that why I suggested to move the code to where it belongs.